pubsub.go

```diff
@@ -213,6 +213,7 @@ const (
 
 type Message struct {
 	*pb.Message
+	ID            string
```

Comment:
This allows msg id to be generated only once. Before it was generated 3 times in a pipeline, which can be expensive when hashing is used. Also, this simplifies code a bit and removes the requirement to MsgIdFunction into MessageCache.

Reply:
Fun fact: One of the [Mina Protocol's motivations](https://github.com/MinaProtocol/mina/blob/rfc/bitswap/rfcs/0042-bitswap.md#motivation) to rework their protocol is high CPU spikes on msg id generation.